[UnifiedPDF] Track text selections on mouse drag

aprotyas opened the pull request:
#### 85b7b3c02480754c7629094106730e68d9c36c7a
<pre>
[UnifiedPDF] Track text selections on mouse drag
<a href="https://bugs.webkit.org/show_bug.cgi?id=268654">https://bugs.webkit.org/show_bug.cgi?id=268654</a>
<a href="https://rdar.apple.com/122200693">rdar://122200693</a>

Reviewed by Tim Horton.

This patch introduces support for tracking text selections on mouse
drag. It does so by keeping a record of the state of selections in the
plugin, through the currentSelection and selectionTrackingData objects.

Selections are initiated on a left button mousedown, and end when a
corresponding mouseup is received.

We additionally provide support for marquee selections, where a
mouse drag gesture with an active alt-key modifier generates a selection
within the rect dragged by the gesture.

Finally, we also add unimplemented codepaths for tracking selections at
different granularities, which is relevant if the user double clicks or
triple clicks, and for extending selections with shift-click.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::annotationForRootViewPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::selectionGranularityForMouseEvent const):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::computeMarqueeSelectionRect):
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::commitCurrentSelection):
(WebKit::UnifiedPDFPlugin::getSelectionString const):

Canonical link: <a href="https://commits.webkit.org/274032@main">https://commits.webkit.org/274032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16473 "Built successfully") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39849 "Hash 53cae5fe for PR 23776 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40117 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33473 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19100 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13635 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31858 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38155 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13874 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12054 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41381 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33965 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34046 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37951 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12635 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10177 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36118 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14072 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8469 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13385 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -154,6 +156,19 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
+        Word,
+        Line,
+    };
+    enum class CommitReason : bool { SelectionIsNoLongerActive, ReceivedMouseUp };
```
UnifiedPDFPlugin::CommitReason : not obvious that it's about selection

aprotyas replied:
Renaming to `SelectionCommitReason`

aprotyas replied:
Renamed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+{
+    if (reason == CommitReason::ReceivedMouseUp && !std::exchange(m_selectionTrackingData.isActive, false))
+        return;
+    LOG(PDF, "Committing text selection: %s", m_currentSelection.get().string.UTF8String);
```
This seems unfortunate privacy-wise

aprotyas replied:
Yep doubly unfortunate because this was only for local debugging. Removing soon

aprotyas replied:
Removed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+        return;
+    LOG(PDF, "Committing text selection: %s", m_currentSelection.get().string.UTF8String);
+    notifySelectionChanged();
+    updateLayerHierarchy();
```
This is unnecessary. I'll remove it.

aprotyas replied:
Fixed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1000,12 +1005,17 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
                 return false;
             }
+
+            beginTrackingSelection(*pageIndex, pointInPageSpace, selectionGranularityForMouseEvent(event), event.modifiers()); // standard text selection
```
Probably remove the comment

aprotyas replied:
Stray comment. Removing soon

aprotyas replied:
Removed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -974,6 +981,8 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
                 return true;
             }
+
+            continueTrackingSelection(*pageIndex, pointInPageSpace);
```
Is there a reason `continueTrackingSelection` implies that the `UnifiedPDFPlugin` could not handle the mouse event for some reason?

aprotyas replied:
No, no, good point. I'll fix the mouse event handler(s) return values.

aprotyas replied:
Returning `true` following [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    if (m_selectionTrackingData.shouldExtendCurrentSelection)
+        notImplemented();
+
+    continueTrackingSelection(pageIndex, pagePoint);
```
Will rewrite this as, since `continueTrackingSelection` does the same check, and there's no reason to bail early here -- we're not doing anything right now anyway.
```suggestion
    // FIXME: <https://webkit.org/b/268617>  Selections should be extensible on shift-click.
    if (m_selectionTrackingData.shouldExtendCurrentSelection)
        notImplemented();

    continueTrackingSelection(pageIndex, pagePoint);
```

aprotyas replied:
Fixed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+
+auto UnifiedPDFPlugin::selectionGranularityForMouseEvent(const WebMouseEvent& event) const -> SelectionGranularity
+{
+    if (event.clickCount() == 2)
+        return SelectionGranularity::Word;
+    if (event.clickCount() == 3)
```
Perhaps some enums for these?

aprotyas replied:
Hmm, I was planning to clean up this code a little more as part of https://webkit.org/b/268616. I'll hold off till then.

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+
+void UnifiedPDFPlugin::setCurrentSelection(RetainPtr<PDFSelection>&& selection)
+{
+    m_currentSelection = selection;
```
Could we add a `WTFMove` here?

aprotyas replied:
Yes!

aprotyas replied:
Fixed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)

aprotyas committed:
[UnifiedPDF] Track text selections on mouse drag
https://bugs.webkit.org/show_bug.cgi?id=268654
rdar://122200693

Reviewed by Tim Horton.

This patch introduces support for tracking text selections on mouse
drag. It does so by keeping a record of the state of selections in the
plugin, through the currentSelection and selectionTrackingData objects.

Selections are initiated on a left button mousedown, and end when a
corresponding mouseup is received.

We additionally provide support for marquee selections, where a
mouse drag gesture with an active alt-key modifier generates a selection
within the rect dragged by the gesture.

Finally, we also add unimplemented codepaths for tracking selections at
different granularities, which is relevant if the user double clicks or
triple clicks, and for extending selections with shift-click.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::annotationForRootViewPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::selectionGranularityForMouseEvent const):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::computeMarqueeSelectionRect):
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::UnifiedPDFPlugin::commitCurrentSelection):
(WebKit::UnifiedPDFPlugin::getSelectionString const):

Canonical link: https://commits.webkit.org/274032@main
